src/CompositeRheologies.jl

```diff
@@ -0,0 +1,794 @@
+) where {T,N,_T}
+    ε  = compute_εII(v, τII, args)
+    return inv(dτII_dεII(v, ε, args))
+end
+
+
+"""
+    dεII_dτII_AD(v::Union{Parallel,CompositeRheology}, τII, args) 
+
+Uses AD to compute the derivative of `εII` vs. `τII`
+"""
+dεII_dτII_AD(v::Union{Parallel,CompositeRheology}, τII, args) = ForwardDiff.derivative(x->compute_εII_AD(v, x, args), τII)
+
+# Computes sum of dεII/dτII for all elements that are NOT parallel elements
+"""
+    dεII_dτII_elements(v::CompositeRheology, TauII, args)
+
+Sums the derivative ∂εII/∂τII (strainrate vs. stress) of all non-parallel elements in a `CompositeRheology` structure. Internally used for jacobian iterations.
+"""
+@inline @generated function dεII_dτII_elements(
+    v::CompositeRheology{T,N}, 
+    TauII::_T, 
+    args
+) where {T, N, _T}
+    quote
+        out = zero(_T)
+        Base.Cartesian.@nexprs $N i ->
+            out += dεII_dτII_nonparallel(v.elements[i], TauII, args)
+    end
+end
+dεII_dτII_nonparallel(v::Any, TauII, args) =   dεII_dτII(v, TauII, args)
+dεII_dτII_nonparallel(v::Parallel, TauII::_T, args) where _T =    zero(_T)
+
+
+
+"""
+    dτII_dεII(v::CompositeRheology, TauII::_T, args)
+
+Computes the derivative of `τII` vs `εII` for `CompositeRheology`   
+"""
+function dτII_dεII(
+    v::CompositeRheology{T,N}, εII::_T, args
+) where {T,N,_T}
+    τ  = compute_τII(v, εII, args)
+    return inv(dεII_dτII(v, τ, args))
+end
+
+
+@generated  function dτII_dεII_i(
+    v::CompositeRheology{T,N}, 
+    εII::_T, 
+    args, I::Int64;
+    tol=1e-6, verbose=false
+) where {T,_T,N}
+    quote
+        Base.@_inline_meta
+        Base.Cartesian.@nexprs $N i -> I == i && return dτII_dεII(v.elements[i], εII, args)
+    end
+end
+
+
+"""
+    dτII_dεII(v::Parallel{T,N}, TauII::_T, args)
+
+Computes the derivative of `τII` vs `εII` for parallel elements   
+"""
+@generated function dτII_dεII(
+    v::Parallel{T,N}, 
+    TauII::_T, 
+    args
+) where {T,N, _T}
+    quote
+        dτII_dεII_der = zero($_T)
+        Base.Cartesian.@nexprs $N i ->
+            dτII_dεII_der += dτII_dεII(v.elements[i], TauII, args)
+        return dτII_dεII_der
+    end
+end
+
+dτII_dεII_AD(v::Union{Parallel,CompositeRheology}, εII, args) = ForwardDiff.derivative(x->compute_τII_AD(v, x, args), εII)
+
+
+
+# HARMONIC AVERAGES (mostly used as initial guesses)
+
+"""
+    compute_τII_harmonic(v::CompositeRheology, EpsII, args)
+
+Harmonic average of stress of all elements in a `CompositeRheology` structure.
+"""
+@inline @generated function compute_τII_harmonic(
+    v::CompositeRheology{T,N}, 
+    EpsII::_T, 
+    args
+) where {T,N, _T}
+    quote
+        out = zero(_T)
+        Base.Cartesian.@nexprs $N i ->
+            out += 1/compute_τII(v.elements[i], EpsII, args)
+        out = 1/out
+    end
+end
+
+
+"""
+    compute_εII_harmonic(v::Parallel{T,N}, TauII::_T, args)
+
+Computes the harmonic average of strainrate for a parallel element
+"""
+@generated function compute_εII_harmonic(
```

Comment:
The function below is identical. Can't they be fused with a `Union`?

Reply:
yes, done